### PR TITLE
Udayan/res filter

### PR DESCRIPTION
**Pull request by Udayan12167**

### CODE REVIEW

**Comment by sils1297** [comment]
CC @fneu for feedback

File: coalib/results/ResultFilter.py [hunk 1]
```diff
@@ -210,15 +211,31 @@ def remove_result_ranges_diffs(result_list, file_dict):
...
 
     :param original_file_dict: Dict of lists of file contents before  changes
     :param modified_file_dict: Dict of lists of file contents after changes
     """
     affected_files = set(original_file_dict.keys()).union(
         set(modified_file_dict.keys()))
+    original_unique_files = affected_files - set(modified_file_dict.keys())
+    renamed_files_dict = {}
     for file in affected_files:
         if file not in original_file_dict:
```
Comment (sils1297) [inline coalib/results/ResultFilter.py]: can you make another commit converting this if into a filter

File: coalib/results/ResultFilter.py [hunk 2]
```diff
@@ -210,15 +211,31 @@ def remove_result_ranges_diffs(result_list, file_dict):
...
-            original_file_dict[file] = []
+            renamed = 0
+            for comparable_file in original_unique_files:
+                s = SequenceMatcher(
+                    None,
+                    ''.join(modified_file_dict[file]),
+                    ''.join(original_file_dict[comparable_file]))
+                if s.real_quick_ratio() >= 0.5:
+                    if s.ratio() > 0.5:
```
Comment (sils1297) [inline coalib/results/ResultFilter.py]: just use `and`, that should be run from left to right. (But confirm that by using your favourite search engine :))

File: coalib/results/ResultFilter.py [hunk 3]
```diff
@@ -210,15 +211,31 @@ def remove_result_ranges_diffs(result_list, file_dict):
...
+                        renamed_files_dict[comparable_file] = file
+                        renamed = 1
+                        break
+            if renamed == 0:
```
Comment (sils1297) [inline coalib/results/ResultFilter.py]: just use `for ... else`, the else clause will be run if you *dont break out of the loop*, this way you can get rid of this `renamed` variable

File: tests/results/ResultFilterTest.py [hunk 4]
```diff
@@ -562,3 +562,84 @@ def test_delete_file_with_result(self):
...
+        tf1_new = abspath('tf1_new')
+
+        original_file_dict = {tf1: testfile_1, tf2: testfile_2}
+        modified_file_dict = {tf1_new: testfile_1}
+
+        renamed_files = ensure_files_present(original_file_dict,
+                                             modified_file_dict)
+
+        self.assertEqual({tf1: tf1_new}, renamed_files)
+
```
Comment (sils1297) [inline tests/results/ResultFilterTest.py]: better make the stuff below an own testcase, they are individual cases aren't they?

File: tests/results/ResultFilterTest.py [hunk 5]
```diff
@@ -562,3 +562,84 @@ def test_delete_file_with_result(self):
...
+        tf2_new = abspath('tf2_new')
+
+        original_file_dict = {tf1: testfile_1, tf2: testfile_2}
+        modified_file_dict = {tf2_new: testfile_2_new}
+
+        renamed_files = ensure_files_present(original_file_dict,
+                                             modified_file_dict)
+
+        self.assertEqual({}, renamed_files)
+
```
Comment (sils1297) [inline tests/results/ResultFilterTest.py]: and so on, make them seperate

**Comment by sils1297** [comment]
CC @fneu for review

File: coalib/results/ResultFilter.py [hunk 6]
```diff
@@ -77,17 +85,22 @@ def basics_match(original_result,
...
 
     :param original_file_dict: Dict of lists of file contents before changes
     :param diff_dict:          Dict of diffs describing the changes per file
     :param original_result_diff_dict: diff for each file for this result
     :param modified_result_diff_dict: guess
+    :param renamed_files:   A dictionary containing file renamings across runs
     :return:                     Boolean value whether the SourceRanges match
     """
     for file_name in original_file_dict:
+        mod_file_name = file_name
```
Comment (sils1297) [inline coalib/results/ResultFilter.py]: put this in else

File: coalib/results/ResultFilter.py (hunk 1 above)
Comment (Udayan12167) [inline coalib/results/ResultFilter.py]: What do you mean by a filter... a separate method?

File: tests/results/ResultFilterTest.py [hunk 7]
```diff
@@ -368,6 +368,288 @@ def test_affected_code(self):
...
+        new_result_list = [res1_pre_change,       # FALSE POSITIVE (in-line)
+                           res1_change,           # correctly kept
+                           res1_post_change,      # FALSE POSITIVE (in-line)
+                           res1_around_change,    # FALSE POSITIVE (in-line)
+                           res1_with_change,      # correctly kept
+                           res1_whole_change,     # correctly kept
+
+                           res1_pre_remove,       # correctly filtered out
+                           res1_post_remove,      # FALSE POSITIVE (in-line)
```
Comment (sils1297) [inline tests/results/ResultFilterTest.py]: ahm did you copy another test? :/ with the false positives? We only want to test renaming here, not everything, right?

File: coalib/results/ResultFilter.py (hunk 1 above)
Comment (sils1297) [inline coalib/results/ResultFilter.py]: `for ... in filter(lambda file: file not in original_file_dict, affected_files):`

File: tests/results/ResultFilterTest.py (hunk 4 above)
Comment (Udayan12167) [inline tests/results/ResultFilterTest.py]: They are just evaluating different conditions of the ensure_files_present. They aren't really different test cases.

Comment (sils1297) [inline tests/results/ResultFilterTest.py]: Disagree. You test certain aspects concerinng file renaming, every aspect should have an own case or it doesn't deserve a test. You're testing several things sequentially here which is bad.

File: tests/results/ResultFilterTest.py (hunk 7 above)
Comment (Udayan12167) [inline tests/results/ResultFilterTest.py]: I wanted to test the full functionality of the ResultFilter with a renamed file. Hence, I reused the test.

Comment (sils1297) [inline tests/results/ResultFilterTest.py]: copying large chunks of code is bad, can we take the other test and give it a parameter? This tests totally unrelated stuff and makes things really complicated. http://pytest.org/latest/parametrize.html#parametrized-test-functions

Comment (Udayan12167) [inline tests/results/ResultFilterTest.py]: http://stackoverflow.com/questions/18182251/does-pytest-parametrized-test-work-with-unittest-class-based-tests

File: tests/results/ResultFilterTest.py (hunk 4 above)
Comment (sils1297) [inline tests/results/ResultFilterTest.py]: this is *so much easier* to understand with every test having an own name describing it's purpose! :)

File: coalib/results/ResultFilter.py [hunk 8]
```diff
@@ -210,15 +211,28 @@ def remove_result_ranges_diffs(result_list, file_dict):
 
 def ensure_files_present(original_file_dict, modified_file_dict):
     """
-    Ensures that all files are available as keys in both dicts.
+    Ensures that all files are available as keys in both dicts. Return a
+    dictionary of renamed files.
```
Comment (sils1297) [inline coalib/results/ResultFilter.py]: we'd rather want a `:return` annotation below the params right?

File: coalib/results/ResultFilter.py [hunk 9]
```diff
@@ -210,15 +211,28 @@ def remove_result_ranges_diffs(result_list, file_dict):
...
 
     :param original_file_dict: Dict of lists of file contents before  changes
     :param modified_file_dict: Dict of lists of file contents after changes
     """
     affected_files = set(original_file_dict.keys()).union(
         set(modified_file_dict.keys()))
+    original_unique_files = affected_files - set(modified_file_dict.keys())
```
Comment (sils1297) [inline coalib/results/ResultFilter.py]: I think it'd be good to do something like:

```
original_files = set(original_file_dict.keys())
modified_files = ...
affected_files = original_files | modified_files
original_unique_files = original_files - modified_files
```

and later instead of `if file not in original_file_dict` you can just `if file not in original_files` as the set might be better optimized for that operation

File: tests/results/ResultFilterTest.py [hunk 10]
```diff
@@ -562,3 +562,87 @@ def test_delete_file_with_result(self):
                                      [old_result_tf1, old_result_tf2],
                                      [new_result])
         self.assertEqual(new_results, [new_result])
+
```
Comment (sils1297) [inline tests/results/ResultFilterTest.py]: can you maybe put all tests that test the `ensure_files_present` in an own class? Then we can rely on them not depending on anything that's needed for the others, less complexity to think about.

File: tests/results/ResultFilterTest.py [hunk 11]
```diff
@@ -562,3 +562,87 @@ def test_delete_file_with_result(self):
...
+        testfile_2_new = ['1\n', '2\n', '0\n', '1\n', '2\n', '1\n', '2\n']
+        tf2_new = abspath('tf2_new')
+
+        original_file_dict = {tf1: testfile_1, tf2: testfile_2}
+        modified_file_dict = {tf2_new: testfile_2_new}
+
+        renamed_files = ensure_files_present(original_file_dict,
+                                             modified_file_dict)
+
+        self.assertEqual({}, renamed_files)
```
Comment (sils1297) [inline tests/results/ResultFilterTest.py]: this commit is awesome!

File: coalib/results/ResultFilter.py [hunk 12]
```diff
@@ -20,12 +20,19 @@ def filter_results(original_file_dict,
...
+                                         modified_file_dict)
     # diffs_dict[file] is a diff between the original and modified file
     diffs_dict = {}
     for file in original_file_dict:
-        diffs_dict[file] = Diff.from_string_arrays(original_file_dict[file],
-                                                   modified_file_dict[file])
+        if file in renamed_files:
+            diffs_dict[file] = Diff.from_string_arrays(
+                original_file_dict[file],
+                modified_file_dict[renamed_files[file]])
```
Comment (sils1297) [inline coalib/results/ResultFilter.py]: why not

```
            diffs_dict[file] = Diff.from_string_arrays(
                original_file_dict[file],
                modified_file_dict[renamed_files.get(file, file)])
```

that's shorter and should yield at least the same performance

File: coalib/results/ResultFilter.py [hunk 13]
```diff
@@ -98,7 +112,7 @@ def source_ranges_match(original_file_dict,
         # original file with file_diff and original_diff applied
         original_total_file = original_total_diff.modified
         # modified file with modified_diff applied
-        modified_total_file = modified_result_diff_dict[file_name].modified
+        modified_total_file = modified_result_diff_dict[mod_file_name].modified
```
Comment (sils1297) [inline coalib/results/ResultFilter.py]: I'd use the same `renamed_files.get(file_name, file_name)` here, no need for the one variable for one use

File: tests/results/ResultFilterTest.py [hunk 14]
```diff
@@ -368,6 +368,72 @@ def test_affected_code(self):
...
+                                               end_line=7,
+                                               end_column=5)
+
+        original_result_list = [res0_pre_change]
+
+        new_result_list = [res1_pre_change,
+                           res1_change,
+                           res1_whole_remove]
+
+        unique_new_result_list = [res1_pre_change,
```
Comment (sils1297) [inline tests/results/ResultFilterTest.py]: shouldn't `res1_pre_change` not be there? It was only in a renamed file, right?

Comment (sils1297) [inline tests/results/ResultFilterTest.py]: ah the line changes, but your testcase doesn't really test this, does it? You should rather add a result that stays the same due to the renaming, hm? Make a result in line 7 (8 in the original file)! That one's shifted and renamed, if we can detect that properly that should work out fine.

File: coalib/results/ResultFilter.py [hunk 15]
```diff
@@ -235,18 +235,21 @@ def ensure_files_present(original_file_dict, modified_file_dict):
...
-                    None,
-                    ''.join(modified_file_dict[file]),
-                    ''.join(original_file_dict[comparable_file]))
-                if s.real_quick_ratio() >= 0.5 and s.ratio() > 0.5:
-                    renamed_files_dict[comparable_file] = file
-                    break
-            else:
-                original_file_dict[file] = []
-        if file not in modified_file_dict:
-            modified_file_dict[file] = []
```
Comment (sils1297) [inline coalib/results/ResultFilter.py]: ouh there are two ifs, yeah drop this commit, that was a bad idea of mine

**Comment by sils1297** [comment]
unack https://github.com/coala-analyzer/coala/pull/2100/commits/292665f40c472014a9e0ec32ee0fbbc4d4b3ab0f and the other commits as well. This also needs a rebase but it's close to merge!

File: tests/results/EnsureFilesPresentTest.py [hunk 16]
```diff
@@ -0,0 +1,94 @@
+import unittest
+from os.path import abspath
+
+from coalib.results.ResultFilter import ensure_files_present
+
+
+class EnsureFilesPresentTest(unittest.TestCase):
+
```
Comment (sils1297) [inline tests/results/EnsureFilesPresentTest.py]: /me feels joy overcoming his body while reading this test class

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala/pull/2100/commits/4aa810c664ec11434608057d4fdd9967152562e0

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala/pull/2100/commits/79b4cb40bfc6fc49032f981828c7dc68296e640e

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala/pull/2100/commits/de5cf0a76531ee6d0f144ebbcda5e2142e65e1c7?diff=split

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala/pull/2100/commits/5dfbe2f43a72445038400c986a0c913ea10b264c

**Comment by sils1297** [comment]
@rultor merge
